### PR TITLE
Contracts: Update Config::Debug

### PR DESCRIPTION
**Pull request by pgherveou**
Update the `Debug` trait that was behind the `unsafe-debug` feature flag, and get rid of the feature flag

Most config should set it to `()` that will simply log contract calls when `runtime::contracts` log target is set to trace.

More advanced debugging tools can leverage this trait to introduce more sophisticated observability behaviors.
See previous PR (#14678) for more context.

Cumulus companion:  https://github.com/paritytech/cumulus/pull/3035

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3414099

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3414325

**Review by pmikolajczyk41 — COMMENTED** [review]
From my point of view there are two separate changes here:

1. removing feature-gating in favor of lightweight default behavior; IMHO that's for better, especially that previously we had many problems with propagating this specific feature flag to CI / general node configuration. However, I'm not fully familiar with Parity's standards/conventions, so I cannot tell whether this is more idiomatic than feature-gating (comparing to e.g. `try-runtime`).

2. moving from a stateless observer to an object-based-callback; to some extent it is a matter of a personal taste, but it has bigger impact; consider a situation where we are interested in just processing outputs from cross-contract calls; previously it was enough to provide a trivial implementation (https://github.com/Cardinal-Cryptography/drink/blob/pallet-breakpoint/drink/src/runtime/debugging.rs); now, it requires building an object that will capture a context before call and its retrieval after call;

I'm not against the change, especially because we are still lacking serious usage of this feature that would prove the optimal API design. Soon, I will try to add a possibility of mocking (intercepting) contract calls, which probably will bring yet another shift in design, so I wouldn't spend too much time now for this.

**Comment by athei** [comment]
I don't think this is a proper replacement for the `Debug` trait which was specifically designed to bundle all future debugging infrastructure without cluttering the  `Config` trait. Hence the generic name and the sub typing.

**Comment by pgherveou** [comment]
> I don't think this is a proper replacement for the `Debug` trait which was specifically designed to bundle all future debugging infrastructure without cluttering the `Config` trait. Hence the generic name and the sub typing.

@athei Is that just a name issue? I can rename it back to Debug if we think its more appropriate.

@pmikolajczyk41 my goal here was to provide the same features that was previously exposed, but provide an implementation that does not add any overhead to the code.

**Comment by athei** [comment]
> @athei Is that just a name issue? I can rename it back to Debug if we think its more appropriate.

This was not a rename but a removal. The `Debug` trait before was just an empty trait to bundle together all the different debugging traits we might want to add. Just as a shortcut name we can mention on the `Config`.

@pmikolajczyk41 I can understand that you want to make the implementations of those traits as simple as possible since you are the one implementing it. But I think you can agree that shifting complexity from on-chain code to a dev tool is the sensible thing to do. Which is exactly what we do by asking the implementor to manage state instead of the pallet.

Regarding conditional compilation: In this case it is viable but as soon as the arguments of the functions require computation we will start causing overhead.

File: frame/contracts/src/debug.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+pub use crate::exec::ExportedFunction;
+use crate::{CodeHash, Config, LOG_TARGET};
+use pallet_contracts_primitives::ExecReturnValue;
+
+/// Umbrella trait for all interfaces that serves for debugging.
+pub trait Debugging<T: Config>: Tracing<T> {}
```
Comment (pgherveou) [inline frame/contracts/src/debug.rs:6]: Debug would be nicer but a bit annoying as it clash with `std:fmt:Debug`, `ConractDebug` would be another good name, any pref?

**Comment by pgherveou** [comment]
@athei  moved back everything under the Debugging umbrella trait, and used back T::Debug associated trait name.

File: frame/contracts/src/debug.rs (hunk 1 above)
Comment (juangirini) [reply frame/contracts/src/debug.rs:6]: `Debugger`?

**Review by juangirini — APPROVED** [review]

File: frame/contracts/src/debug.rs (hunk 1 above)
Comment (athei) [reply frame/contracts/src/debug.rs:6]: I like `Debugger`

**Review by athei — APPROVED** [review]

File: frame/contracts/src/debug.rs (hunk 1 above)
Comment (pgherveou) [reply frame/contracts/src/debug.rs:6]: Will update and follow consensus

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3429539

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3429540

**Comment by pgherveou** [comment]
@oleg-plakida  you mind giving us a +1 here for removing the `unsafe-debug` feature added in the previous PR #14678

**Comment by pgherveou** [comment]
> @oleg-plakida you mind giving us a +1 here for removing the `unsafe-debug` feature added in the previous PR #14678

cc @rcny

**Review by alvicsam — APPROVED** [review]

File: bin/node/runtime/Cargo.toml [hunk 2]
```diff
@@ -372,5 +372,4 @@ try-runtime = [
 	"pallet-vesting/try-runtime",
 	"pallet-whitelist/try-runtime",
 	"sp-runtime/try-runtime",
-]
-unsafe-debug = [ "pallet-contracts/unsafe-debug" ]
+]
```
Comment (pgherveou) [inline bin/node/runtime/Cargo.toml:375]:
```suggestion
]

```

**Review by sergejparity — APPROVED** [review]

**Comment by pgherveou** [comment]
bot merge

**Pull request merged**